# **Day10**

Created PythonTutorial database in using mysql command line

pip install mysql-connector-python

## **Database Handling**

In [1]:
# how to create database using Python

import mysql.connector as m
 
mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64")
cursor=mydatabase.cursor()
cursor.execute("CREATE DATABASE PYTHONTUTORIAL")   # go and check inside MySQL whether database has been created or not


eX2

In [2]:
# how to create table using Python

import mysql.connector as m
mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64",database="PYTHONTUTORIAL")

cursor=mydatabase.cursor()

cursor.execute("create table dept(deptno int primary key auto_increment,dname varchar(20),loc varchar(30))")

Ex3

In [3]:
# insert records in the table

mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64",database="PYTHONTUTORIAL")

cursor=mydatabase.cursor()

query1="insert into dept(dname,loc) values(%s,%s)"    # it has to be "s"
mydata1=('accounts','delhi')
cursor.execute(query1,mydata1)
mydatabase.commit()                    #   commit() is compulsory

Ex4

In [7]:
# accept values from user and then insert the record
# run this program on "terminal"


mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64",database="PYTHONTUTORIAL")


query="insert into dept(dname,loc) values(%s,%s)"         #  must be "s"
dname=input("enter department name")
loc=input("enter location")
cursor=mydatabase.cursor()
cursor.execute(query,[dname,loc])       #  second argument has to be list or tuple or dictionary
mydatabase.commit()      #   commit() is must here

Ex5

In [8]:
mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64",database="PYTHONTUTORIAL")


cursor=mydatabase.cursor()
query="insert into dept(dname,loc) values(%s,%s)"
cursor.execute(query,['admin','Banglore'])
cursor.execute(query,['RnD','Chennai'])
mydatabase.commit()

Ex6


In [11]:
mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64",database="PYTHONTUTORIAL")

cursor=mydatabase.cursor()
query="insert into dept(dname,loc) values(%s,%s)"
while True:
    dname=input("Enter dname")
    loc=input("Enter location")
    cursor.execute(query,[dname,loc])
    ans=input("Do you wish to continue")
    if(ans=="no"):
        break
mydatabase.commit()

In [12]:
print(type(cursor))

<class 'mysql.connector.cursor_cext.CMySQLCursor'>


Ex7

In [13]:
mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64",database="PYTHONTUTORIAL")

cursor=mydatabase.cursor()

query="update dept set loc=%s where deptno=%s"
loc=input("Enter location")
deptno=input("Enter deptno")
cursor.execute(query,[loc,deptno])
mydatabase.commit()

Ex8

In [14]:
mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64",database="PYTHONTUTORIAL")

cursor=mydatabase.cursor()
query="delete from dept where dname=%s"

dname=input("Enter department to delete the record")
cursor.execute(query,(dname,))     # passing tuple as second argument   comma(,)  is mandatory here as we just have only one value inside the tuple

mydatabase.commit();

Ex9

In [15]:
# executemany function  if we pass list of list to the "execute()" method

from dataclasses import dataclass
import mysql.connector as m
mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64",database="PYTHONTUTORIAL")

cursor=mydatabase.cursor()
query="insert into dept(dname,loc) values(%s,%s)"
mylist=[['HR','Mumbai'],['Placement','Pune']]      # list of list
# cursor.execute(query,mylist)    #  won't work as it is list of list
cursor.executemany(query,mylist)
mydatabase.commit()

Ex10

In [16]:
import mysql.connector as m
mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64",database="PYTHONTUTORIAL")

cursor=mydatabase.cursor()
query="select * from dept"
cursor.execute(query)
result=cursor.fetchall()     # here we get tuples equivalent to the number of records
for record in result:
    print(record)


(1, 'accounts', 'Liberty City')
(2, 'Technology', 'California')
(4, 'RnD', 'Chennai')
(5, 'Science', 'Texas')
(6, 'Finance', 'Germany')
(7, '', '')
(8, '', '')
(9, '', '')
(10, '', '')
(11, 'HR', 'Mumbai')
(12, 'Placement', 'Pune')


Ex11

In [17]:
import mysql.connector as m
with m.connect(host="localhost",user="root",password="Mysqlroot@64",database="PYTHONTUTORIAL") as mydatabase:
    cursor=mydatabase.cursor()
    query="select * from dept"
    cursor.execute(query)
    result=cursor.fetchall()     # here we get tuples equivalent to the number of records
    for record in result:
      print(record)


(1, 'accounts', 'Liberty City')
(2, 'Technology', 'California')
(4, 'RnD', 'Chennai')
(5, 'Science', 'Texas')
(6, 'Finance', 'Germany')
(7, '', '')
(8, '', '')
(9, '', '')
(10, '', '')
(11, 'HR', 'Mumbai')
(12, 'Placement', 'Pune')


Ex11

In [18]:
import mysql.connector as m
mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64",database="PYTHONTUTORIAL")

cursor=mydatabase.cursor()
query="select * from dept"
cursor.execute(query)
result=cursor.fetchall()     # here we get tuples equivalent to the number of records
for record in result:
    for i in range(0,len(record)):    # traversing through each and every column of respective record
        print(record[i])

1
accounts
Liberty City
2
Technology
California
4
RnD
Chennai
5
Science
Texas
6
Finance
Germany
7


8


9


10


11
HR
Mumbai
12
Placement
Pune


Ex12

In [19]:
import mysql.connector as m

mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64",database="PYTHONTUTORIAL")
cursor=mydatabase.cursor()

query="insertt into dept(dname,loc) values(%s,%s)"
try:
    cursor.execute(query,['training','Pune'])
    cursor.execute(query,['cleaning','Banglore'])
    mydatabase.commit()
except Exception as e:
    print("exception occured\t",e)
    mydatabase.rollback()
print("done")

exception occured	 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'insertt into dept(dname,loc) values('training','Pune')' at line 1
done


In [20]:
#correct code for upper error 

import mysql.connector as m

mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64",database="PYTHONTUTORIAL")
cursor=mydatabase.cursor()

query="insert into dept(dname,loc) values(%s,%s)"
try:
    cursor.execute(query,['Training','Pune'])
    cursor.execute(query,['Sodexo','Banglore'])
    mydatabase.commit()
except Exception as e:
    print("exception occured\t",e)
    mydatabase.rollback()
print("done")

done


Ex13

In [26]:
# read from csv and insert into database table

import pandas as pd
import mysql.connector as m

empdata = pd.read_csv('t_csv.csv', index_col=False, delimiter = ',')

mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64",database="PYTHONTUTORIAL")

cursor=mydatabase.cursor()

cursor.execute('DROP TABLE IF EXISTS book;')
cursor.execute("create table book(id int(3), name varchar(20))")

for i,row in empdata.iterrows():
            #here %S means string values 
            sql = "INSERT INTO book VALUES (%s,%s)"
            cursor.execute(sql, tuple(row))
            # the connection is not auto committed by default, so we must commit to save our changes

mydatabase.commit()
print("Records inserted")


Records inserted


Ex14

In [27]:
# read from database and store inside CSV

import pandas as pd
import mysql.connector as m

mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64",database="PYTHONTUTORIAL")

cursor=mydatabase.cursor()

query="select id, name from book"
cursor.execute(query)
result=cursor.fetchall()     # here we get tuples equivalent to the number of records
names=[]
ids=[]

for id,name in result:
    ids.append(name)
    names.append(name)
    # addresses.append(address)

mydictionary={'Ids':ids,'Name':names}
#print(mydictionary)

df=pd.DataFrame(mydictionary)
df_csv=df.to_csv('NewBook.csv',index=False)



## **Socket Prgm**

Ex1

In [ ]:


# bind() method  binds the server to a specific IP and port so that it can listen to incoming requests on that IP and port.
# listen() method puts the server into listen mode
# accept() method allows server to wait for client to send the request

""" The process of converting string objects to byte objects is called encoding and the inverse is called decoding. 

The bytes() method takes in an object (a string in our case), the required encoding method, and convert it into a byte object. 

 """
import socket
with socket.socket() as s:
    print("socket created")
    s.bind(("localhost",5000))
    s.listen()
    print("waiting for the client")

    client,address=s.accept()  # returns tuple[socket, _RetAddress] 
    print("client is\t",client)
    print("address is\t",address)
    # converts byte string received from client
    # into string format
    name=client.recv(1024).decode()
    print("connected with\t",address,name)

    # converting the string into bytes using the
    # UTF-8 encoding.
    client.send(bytes(f"{name} welcome to the server side",'utf-8'))
    
    

socket created
waiting for the client


Run server first then go for client 

In [ ]:
import socket

with socket.socket() as c:
    c.connect(("localhost",5000))
    name=input("Enter your name")

    # convert the input string into bytes.
    c.send(bytes(name,'utf-8'))

    # decodes the response which comes as byte string
    # into a standard string format
    print(c.recv(1024).decode())


Ex2

In [ ]:
# bind() method  binds the server to a specific IP and port so that it can listen to incoming requests on that IP and port.
# listen() method puts the server into listen mode
# accept() method allows server to wait for client to send the request

""" The process of converting string objects to byte objects is called encoding and the inverse is called decoding. 

The bytes() method takes in an object (a string in our case), the required encoding method, and convert it into a byte object. 

 """
import socket
with socket.socket() as s:
    print("socket created")
    s.bind(("localhost",5000))
    s.listen()
    print("waiting for the client")

    client,address=s.accept()  # returns tuple[socket, _RetAddress] 
    print("client is\t",client)
    print("address is\t",address)
    # converts byte string received from client
    # into string format
    name=client.recv(1024).decode()
    print("connected with\t",address,name)

    # converting the string into bytes using the
    # UTF-8 encoding.
    client.send(bytes(f"{name} welcome to the server side",'utf-8'))
    
    

In [ ]:
import socket

with socket.socket() as c:
    c.connect(("localhost",5000))
    name=input("Enter your name")

    # convert the input string into bytes.
    c.send(bytes(name,'utf-8'))

    # decodes the response which comes as byte string
    # into a standard string format
    print(c.recv(1024).decode())

## **Pickling_Unpicklikng**

Ex1

In [4]:
import pickle

class Student:
    def __init__(self,name,age):
        self.name=name
        self.age=age
    def __str__(self):
        return ""+self.name+"\t"+str(self.age)

s1=Student("Rohit",25)
print("After creating object")
print(s1)
print(id(s1))

# pickling or serialization
pickle.dump(s1,open("mystudent.pkl","wb"))


s1=None
print("After storing object and after making s1 None")
print(s1)
print(id(s1))

print("Not to worry, we've stored object inside the filesystem, so it can be retrieved")

# unpickling or deserialization
temp=pickle.load(open("mystudent.pkl","rb"))
print(temp)

print(id(temp))

After creating object
Rohit	25
2209238528656
After storing object and after making s1 None
None
140703254416144
Not to worry, we've stored object inside the filesystem, so it can be retrieved
Rohit	25
2209238430864


Ex2



In [5]:
# Pickling Unpickling
# Pickling means storing an entire object inside the file system ( a.k.a.  Serialization)
# Unpickling meand retrieving an entire object from the file system ( a.k.a.  Deserialization)

import pickle
import pandas as pd


df=pd.read_csv("indices_close.csv")

# serialization or pickling
pickle.dump(df,open("indices_close.pkl","wb"))

# deserialization or unpickling
mydf=pickle.load(open("indices_close.pkl","rb"))
print("displaying dataframe which just has been read")
print(mydf.head())

displaying dataframe which just has been read
      Index Name  Index Date Open Index Value High Index Value  \
0       Nifty 50  05-08-2024         24302.85         24350.05   
1  Nifty Next 50  05-08-2024         71459.55         72084.95   
2      Nifty 100  05-08-2024         25291.05         25375.65   
3      Nifty 200  05-08-2024         13731.25          13786.2   
4      Nifty 500  05-08-2024         22772.55         22868.35   

  Low Index Value  Closing Index Value  Points Change Change(%)      Volume  \
0         23893.7             24055.60        -662.10     -2.68   486994718   
1        70269.45             70936.90       -2391.25     -3.26   775888423   
2        24871.55             25043.05        -728.85     -2.83  1267257581   
3           13508             13600.20        -412.70     -2.95  3131433955   
4         22397.7             22542.45        -717.00     -3.08  4019339380   

  Turnover (Rs. Cr.)    P/E   P/B Div Yield  
0           45066.65  22.38  4.05   

## **How to Create PDF**

In [22]:
# 1) install "fpdf"

# pip install fpdf

# 2) 
# now write following code:

from fpdf import FPDF

ref=FPDF()
ref.add_page()
ref.set_font("Arial",size=20)
ref.cell(200,95,"Welcome to PDF world",align="C", fill=0, border=1)

ref.output("pythonpdf.pdf")

''

## **How to use WhatsApp using Python**

Open powershell window and give following command:

pip install pywhatkit

Type the following code in the python file:

	First.py

import pywhatkit
# here 7,52 is time on which you want whatsapp message to go to the receiver

pywhatkit.sendwhatmsg('+919270682272','Good Morning boss',7,52)


	Run the above code , it will give you message:

In 94 Seconds WhatsApp will open and after 15 Seconds Message will be Delivered!

open the browser and wait
	automatically "whatsapp web" will start and you can see the message.

Also it will create "PyWhatKit_DB" file in the current working directory.



Date: 31/10/2022
Time: 7:50
Phone Number: +919270682272
Message: Good Morning boss
--------------------


In [3]:
import pywhatkit

# here 7,52 is time on which you want whatsapp message to go to the receiver

pywhatkit.sendwhatmsg('+918355841252','Good Morning Hemant',10,29)

In 33 Seconds WhatsApp will open and after 15 Seconds Message will be Delivered!


## **GUI Prgm**

Tkinter is a standard Python GUI library which gives us an object oriented interface to the Tk GUI toolkit.

how to install?

pip install tk

In [4]:
# mysql> create table person(id int(4) primary key auto_increment, name varchar(30),address varchar(40),age int(4));

# create on mysql cmd line 

Ex1

In [10]:
# window with button

import tkinter

mywindow=tkinter.Tk()
mywindow.title("My Window")
mywindow.geometry("500x200")
mybutton=tkinter.Button(mywindow,text="ok",command=mywindow.destroy).pack(side='top')
mywindow.mainloop()

"""
window.mainloop() tells Python to run the Tkinter event loop. This method listens for events, such as button clicks or keypresses,
 and blocks any code that comes after it from running until you close the window where you called the method.
"""

'\nwindow.mainloop() tells Python to run the Tkinter event loop. This method listens for events, such as button clicks or keypresses,\n and blocks any code that comes after it from running until you close the window where you called the method.\n'

Ex2

In [13]:
from tkinter import *
import tkinter.font as font

gui = Tk(className='Python Examples - Button')
gui.geometry("500x200")

# define font
myFont = font.Font(family='Helvetica')

# create button
button = Button(gui, text='My Button', bg='black', fg='red',font=myFont)
# apply font to the button label
#button['font'] = myFont
# add button to gui window
button.pack()

gui.mainloop() 

Ex3

In [14]:
from tkinter import *
import tkinter.font as font

def myfun():
    text=button['text']
    if(text=="ok"):
        button['text']="cancel"
    else:
        button['text']="ok"


gui = Tk(className='Python Examples - Button')
gui.geometry("500x200")

# define font
myFont = font.Font(family='Helvetica')

# create button
button = Button(gui, text='ok', bg='black', fg='red',font=myFont,command=myfun)
# apply font to the button label
#button['font'] = myFont
# add button to gui window
button.pack()

gui.mainloop()

Ex4

In [15]:
# Change Button Background Color during Mouse Click

from tkinter import *
import tkinter.font as font

gui = Tk(className='Python Examples - Button')
gui.geometry("500x200")

# define font
myFont = font.Font(family='Helvetica')

# create button
button = Button(gui, text='ok', bg='green', activebackground='red',font=myFont)
# apply font to the button label
#button['font'] = myFont
# add button to gui window
button.pack()

gui.mainloop() 

Ex5

In [22]:
from tkinter import *


def printDetails() :
    usernameText = usernameEntry.get()
    finaltext="user entered :"+usernameText
    displaylabel['text']=finaltext
    

#window
tkWindow = Tk()  
tkWindow.geometry('400x150')  
tkWindow.title('Python Examples')

#label
usernameLabel = Label(tkWindow, text="Enter your name")
#entry for user input
usernameEntry = Entry(tkWindow)

displaylabel=Label(tkWindow,text="")


#submit button
submitButton = Button(tkWindow, text="Submit", command=printDetails)

#place label, entry, and button in grid
usernameLabel.grid(row=0, column=0)
usernameEntry.grid(row=0, column=1) 
submitButton .grid(row=1, column=1)  
displaylabel.grid(row=5,column=1)

#main loop
tkWindow.mainloop()

Ex6

In [25]:

from tkinter import *
from functools import partial
import tkinter.font as font

def validateLogin():
    myFont = font.Font(family='Helvetica',size=20)
    if(usernameEntry.get()=="scott" and passwordEntry.get()=="tiger"):
        resultlabel.config(fg="green",font=myFont)
        resultlabel["text"]="Valid User"
    else:
        resultlabel.config(fg="red",font=myFont)
        resultlabel["text"]="Invalid User"
	

#window
tkWindow = Tk()  
tkWindow.geometry('400x150')  
tkWindow.title('Tkinter Login Form - pythonexamples.org')

#username label and text entry box
usernameLabel = Label(tkWindow, text="User Name")
usernameLabel.grid(row=0, column=0)

usernameEntry = Entry(tkWindow)
usernameEntry.grid(row=0, column=1)   

#password label and password entry box
passwordLabel = Label(tkWindow,text="Password")
passwordLabel.grid(row=1, column=0)  

passwordEntry = Entry(tkWindow, show='*')
passwordEntry.grid(row=1, column=1)  

resultlabel=Label(tkWindow)
resultlabel.grid(row=6,column=0)


#login button
loginButton = Button(tkWindow, text="Login", command=validateLogin)
loginButton.grid(row=4, column=0)  



tkWindow.mainloop()

Ex7

In [28]:
# TKinter and database


from tkinter import *
from functools import partial
import tkinter.font as font


import mysql.connector as m

mydatabase=m.connect(host="localhost",user="root",password="Mysqlroot@64",database="pythontutorial")
query="insert into person(id,name) values(%s,%s)"         #  must be "s"


def savePerson():
    myFont = font.Font(family='Helvetica',size=20)
    name_1=nameEntry.get()
    id_1=idEntry.get()
    age_1=ageEntry.get()
    #print(name_1,address_1,age_1)
    cursor=mydatabase.cursor()
    cursor.execute(query,[id_1,name_1])       #  second argument has to be list or tuple or dictionary
    mydatabase.commit()
    

#window
tkWindow = Tk()  
tkWindow.geometry('400x150')  
tkWindow.title('Tkinter Login Form - pythonexamples.org')

#name label and text entry box
nameLabel = Label(tkWindow, text="Name")
nameLabel.grid(row=0, column=0)
nameEntry = Entry(tkWindow)
nameEntry.grid(row=0, column=1)  

#Address label and password entry box
idLabel = Label(tkWindow,text="Address")
idLabel.grid(row=1, column=0)  

idEntry = Entry(tkWindow)
idEntry.grid(row=1, column=1)  

#Age label and password entry box
# ageLabel = Label(tkWindow,text="Age")
# ageLabel.grid(row=2, column=0)  

# ageEntry = Entry(tkWindow)
# ageEntry.grid(row=2, column=1)  


#login button
saveButton = Button(tkWindow, text="Save", command=savePerson)
saveButton.grid(row=4, column=0)  

tkWindow.mainloop()


Exception in Tkinter callback
Traceback (most recent call last):
  File "c:\Users\RKC\anaconda3\Lib\tkinter\__init__.py", line 1967, in __call__
    return self.func(*args)
           ^^^^^^^^^^^^^^^^
  File "C:\Users\RKC\AppData\Local\Temp\ipykernel_2348\3052028860.py", line 19, in savePerson
    age_1=ageEntry.get()
          ^^^^^^^^^^^^^^
  File "c:\Users\RKC\anaconda3\Lib\tkinter\__init__.py", line 3131, in get
    return self.tk.call(self._w, 'get')
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_tkinter.TclError: invalid command name ".!entry3"


## **Virtual Env**

>A virtual environment is a tool that helps to keep dependencies required by different projects separate by creating isolated python virtual environments for them. This is one of the most important tools that most Python developers use.

>Why do we need a virtual environment?

>>Imagine a scenario where you are working on two web-based python projects one of them uses Django 4.0 and the other uses Django 4.1. In such situations virtual environment can be really useful to maintain the dependencies of both projects.

>>By default, every project on your system will use these same directories to store and retrieve site packages (third-party libraries). How does this matter? Now, in the above example of two projects, you have two versions of Django. This is a real problem for Python since it can’t differentiate between versions in the “site-packages” directory. So both v1.9 and v1.10 would reside in the same directory with the same name. This is where virtual environments come into play. To solve this problem, we just need to create two separate virtual environments for both projects. The great thing about this is that there are no limits to the number of environments you can have since they’re just directories containing a few scripts. A virtual Environment should be used whenever you work on any Python-based project. It is generally good to have one new virtual environment for every Python-based project you work on. So the dependencies of every project are isolated from the system and each other.




your project location should be inside 

	-------------------------------



main location of "site-packages" where all libraries are by default downloaded is:


"C:\Users\RKC\AppData\Local\Programs\Python\Python311\Lib\site-packages" - 146 packages

"C:\Users\RKC\anaconda3\Lib\site-packages" - 1227 packages


if you have installed a package and still you get package not found error, always select "existing" as an option and select interpreter from 
varia
**"C:\Users\RKC\anaconda3\python.exe"**

How to create virtual environment?

first install virtualenv

[ virtualenv is used to manage Python packages for different projects. Using virtualenv allows you to avoid installing Python packages globally which could break system tools or other projects.]



**pip install virtualenv**



if you want to create a virtual environment , go to your project location on the terminal and type:

>C:\Users\Sriram\PycharmProjects> **python -m venv myenv**

>>this command will create "myenv" folder inside "C:\Users\Sriram\PycharmProjects"
>>>inside "myenv" folder    you will get
>>>>"Lib\site-packages"
>>>>>this "site-packages" is the folder where your environment (myenv) specific packages only will be stored. 
i.e. [C:\Users\Sriram\PycharmProjects\myenv\Lib\site-packages]

>>>>>>so, the "site-packages" folder which is available to all the projects
	i.e.
>>>>>>C:\Users\Sriram\AppData\Local\Programs\Python\Python310\Lib\site-packages

is different and "site-packages" which is specific to a particular virtual environment  i.e.
>>>>>>>>C:\Users\Sriram\PycharmProjects\myenv\Lib\site-packages



if you want to list out the packages installed for your project, go to your project folder on the terminal and give following command

C:\Users\Sriram\PycharmProjects>pip list

it will list out all the packages which have been installed globally.


now if you want to activate your virtual environment, give following command

C:\Users\Sriram\PycharmProjects> cd myenv
it will change to 
C:\Users\Sriram\PycharmProjects\myenv>

now give command
C:\Users\Sriram\PycharmProjects\myenv> Scripts\activate

make sure command prompt changes to :

(myenv) C:\Users\Sriram\PycharmProjects\myenv>
	this means that we've activated "myenv" virtual environment

now list out how many packages have been installed inside your virtual environment (myenv)

(myenv) C:\Users\Sriram\PycharmProjects\myenv> pip list


